### DISCUSSION TITLE
Using the Ghidra API with Datatype archives and Comments

### DISCUSSION THREAD

**Discussion by Nemoumbra** [NationalSecurityAgency/ghidra#5620] [Q&A] [answered]
The API is well-documented, but it's very hard to get the initial understanding of where I'm supposed to acquire certain objects for the function arguments. At the same time it's hard to figure out how exactly one is supposed to approach their problem.
Now let's get to the point.
`currentProgram.getDataTypeManager()` can be used to get the archive associated with... well... the currently open program. Then I can iterate over the needed data types with methods `.getAllFunctionDefinitions()`, etc.
But I've opened some archives in the Data Type Manager window and now I want to acquire a "handle" for them... like, iterate over all DataTypeManagers or request one by name. There is a way to open an archive file manually with `openArchive(java.io.File, boolean)` or to open a standard one (contained within the Ghidra installation), but it's kind of not it.

Next, is there a good way to iterate over comments in the whole program? I've found something very similar to this, but one piece of a puzzle is still missing:
```
listing = currentProgram.getListing()
# Get the argument for the method "getCommentAddressIterator"
address_set = currentProgram.getAddressFactory().getAddressSet()

for addr in listing.getCommentAddressIterator(address_set, True):
    comment = listing.getComment(<comment_type>, addr)
    do_stuff(comment)
```
This <comment_type> is supposed to be a static field of the class `CodeUnit`, for instance, `CodeUnit.REPEATABLE_COMMENT` (but these are just ints so we can use 0,1,2, etc). Is there no better way to get all available comments at an arbitrary address other than calling this 5 or so times with different args? It certainly doesn't look like something we can easily extend without manually adding the new item if a new comment type is introduced.

**Comment by ryanmkurtz** [comment]
There is a `CommentFieldSearcher` class that that the GUI uses, but it looks like even that requires a separate instance for each comment type.  I also cannot find a way to avoid the repeated calls.

**Comment by dragonmacher** [comment, marked as answer]
For working with DataTypeManagers, you can use a service to do that, for example, this method is taken from the `PrintStructureScript`:

```
	private DataType findDataTypeByName(String name) {
		PluginTool tool = state.getTool();
		DataTypeManagerService service = tool.getService(DataTypeManagerService.class);
		DataTypeManager[] dataTypeManagers = service.getDataTypeManagers();
		for (DataTypeManager manager : dataTypeManagers) {
			DataType dataType = manager.getDataType(name);
			if (dataType != null) {
				return dataType;
			}
		}
		return null;
	}
```

This shows how to get and use the `DataTypeManagerService`.  Note this service implements the `DataTypeArchiveService`, which has methods you may want to use.

**Comment by dragonmacher** [comment]
Regarding getting all comments in a program, this is not something that is part of the normal Ghidra workflow, thus, there is no existing method to do it.  I don't believe we have added a new comment type in decades, so you should not need to change your code often once written.

**Reply by Nemoumbra** [reply]
I see, thank you!
I've got another question, by the way, maybe you know the answer to this too?
How can I check if an arbitrary address is valid (which means it will be displayed in red in the listing)?
![image](https://github.com/NationalSecurityAgency/ghidra/assets/62447396/1d49c0ac-6820-4cf8-b7db-e1d8fac450da)
I've grabbed the pointer via `ptr = getDataAt(<address>).getValue()`, now all what's left is to do something like `ptr.isValidAddress()`. I don't see anything like that in the docs. The standard space (ram:) is very greedy and says that the range from `ptr` to `ptr+1` is valid.... It also says that nullptr is a valid address too...
I tried to look for the "Address not in memory" tooltip, but it looks like there the devs make a call to a method for generating the preview of the location and if it fails, the address is declared to be invalid. That's convenient for them, but I won't bring such heavy guns to my script and I doubt this is even close to how it needs to be done.

**Reply by dragonmacher** [reply]
You should be able to check if the address is in memory like this:
```
Memory memory = currentProgram.getMemory();
if (!memory.contains(address)) {
```
A symbol will be red if it is not in memory or if it is an external address/symbol that is not linked.
